File: flink-java-code/src/main/java/com/atguigu/day05/GenPeriodicWatermark.java
package com.atguigu.day05;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.streaming.api.watermark.Watermark;
import org.apache.flink.util.Collector;

import javax.annotation.Nullable;

public class GenPeriodicWatermark {
    public static void main(String[] args) throws Exception{
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.getConfig().setAutoWatermarkInterval(10*1000L);
        DataStream<String> stream = env.socketTextStream("localhost", 9999, '\n');
stream.map(new MapFunction<String, Tuple2<String,Long>>() {
    @Override
    public Tuple2<String, Long> map(String s) throws Exception {
        String[] arr = s.split(" ");
        return Tuple2.of(arr[0],Long.parseLong(arr[1])*1000L);
    }
}).assignTimestampsAndWatermarks(new AssignerWithPeriodicWatermarks<Tuple2<String, Long>>() {
    private final Long bound = 10 * 1000L;
    private Long maxTs = Long.MIN_VALUE + bound +1;
    @Override
    public Watermark getCurrentWatermark() {
        System.out.println("generate watermark: " + (maxTs - bound - 1) + "ms");
        return new Watermark(maxTs - bound - 1);
    }

    @Override
    public long extractTimestamp(Tuple2<String, Long> element, long recordTimestamp) {
        System.out.println("extract timestamp!!!");
        maxTs = Math.max(maxTs, element.f1);
        return element.f1;
    }
}).keyBy(r ->r.f0).process(new KeyedProcessFunction<String, Tuple2<String, Long>, String>() {
    @Override
    public void processElement(Tuple2<String, Long> value, Context ctx, Collector<String> out) throws Exception {
        out.collect("current watermark is: " + ctx.timerService().currentWatermark());
    }
}).print();

env.execute();

    }
}

File: FlinkTutorial/src/main/scala/com/atguigu/day06/GenPeriodicWatermark.scala
package com.atguigu.day06

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.{AssignerWithPeriodicWatermarks, KeyedProcessFunction, ProcessFunction}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.watermark.Watermark
import org.apache.flink.util.Collector

object GenPeriodicWatermark {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    env.getConfig.setAutoWatermarkInterval(10*1000L)

    val stream: DataStream[String] = env.socketTextStream("localhost",9999,'\n')

    stream.map(line =>{
      val arr: Array[String] = line.split(" ")
      (arr(0),arr(1).toLong *1000L)
    }).assignTimestampsAndWatermarks(
      new AssignerWithPeriodicWatermarks[(String, Long)] {
        val bound = 10 * 1000L // 最大延迟时间
        var maxTs = Long.MinValue + bound // 当前观察到的最大时间戳
        // 用来生成水位线
        // 默认200ms调用一次
        override def getCurrentWatermark: Watermark = {
          println("generate watermark!!!" + (maxTs - bound - 1) + "ms")
          new Watermark(maxTs - bound - 1)
        }
        // 每来一条数据都会调用一次
        override def extractTimestamp(element: (String, Long), recordTimestamp: Long): Long = {
          println("extract timestamp!!!")
          maxTs = maxTs.max(element._2) // 更新观察到的最大事件时间
          element._2 // 抽取时间戳
        }
      }
    ).keyBy(r => r._1).process(new KeyedProcessFunction[String,(String,Long),String] {
      override def processElement(value: (String, Long), ctx: KeyedProcessFunction[String, (String, Long), String]#Context, out: Collector[String]): Unit = {
        out.collect("current watermark is " + ctx.timerService().currentWatermark())
      }
    }
    )
    env.execute()
  }
}
